File: frontend/lib/domain/entities/verification_checkpoint_config.dart
import 'package:equatable/equatable.dart';

enum CheckpointType {
  boolean,
  text,
  textarea,
  rating,
  image,
  video,
  multiImage,
  dropdown,
  number;

  static CheckpointType fromString(String value) {
    switch (value.toLowerCase()) {
      case 'boolean':
        return CheckpointType.boolean;
      case 'text':
        return CheckpointType.text;
      case 'textarea':
        return CheckpointType.textarea;
      case 'rating':
        return CheckpointType.rating;
      case 'image':
        return CheckpointType.image;
      case 'video':
        return CheckpointType.video;
      case 'multi_image':
        return CheckpointType.multiImage;
      case 'dropdown':
        return CheckpointType.dropdown;
      case 'number':
        return CheckpointType.number;
      default:
        return CheckpointType.text;
    }
  }

  String get value {
    switch (this) {
      case CheckpointType.boolean:
        return 'boolean';
      case CheckpointType.text:
        return 'text';
      case CheckpointType.textarea:
        return 'textarea';
      case CheckpointType.rating:
        return 'rating';
      case CheckpointType.image:
        return 'image';
      case CheckpointType.video:
        return 'video';
      case CheckpointType.multiImage:
        return 'multi_image';
      case CheckpointType.dropdown:
        return 'dropdown';
      case CheckpointType.number:
        return 'number';
    }
  }
}

class VerificationCheckpointConfigEntity extends Equatable {
  final String id;
  final String checkpointKey;
  final String label;
  final CheckpointType checkpointType;
  final String? description;
  final bool isRequired;
  final bool isEnabled;
  final int displayOrder;
  final Map<String, dynamic>? validation;
  final List<String>? options;
  final String? placeholder;
  final String? hint;
  final String? category;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  const VerificationCheckpointConfigEntity({
    required this.id,
    required this.checkpointKey,
    required this.label,
    required this.checkpointType,
    this.description,
    this.isRequired = true,
    this.isEnabled = true,
    this.displayOrder = 0,
    this.validation,
    this.options,
    this.placeholder,
    this.hint,
    this.category,
    this.createdAt,
    this.updatedAt,
  });

  @override
  List<Object?> get props => [
        id,
        checkpointKey,
        label,
        checkpointType,
        description,
        isRequired,
        isEnabled,
        displayOrder,
        validation,
        options,
        placeholder,
        hint,
        category,
        createdAt,
        updatedAt,
      ];
}

File: backend_python/app/models/verification_checkpoint_config.py
"""
Verification Checkpoint Configuration model
Defines what details nodal officers need to verify and submit
"""
from datetime import datetime
from typing import Optional, Dict, Any, List
from beanie import Document
from pydantic import Field, field_validator
from enum import Enum


class CheckpointType(str, Enum):
    """Types of verification checkpoints"""
    BOOLEAN = "boolean"  # Yes/No question
    TEXT = "text"  # Text feedback
    TEXTAREA = "textarea"  # Long text feedback
    RATING = "rating"  # 1-5 star rating
    IMAGE = "image"  # Image upload
    VIDEO = "video"  # Video upload
    MULTI_IMAGE = "multi_image"  # Multiple images
    DROPDOWN = "dropdown"  # Dropdown selection
    NUMBER = "number"  # Numeric value


class VerificationCheckpointConfig(Document):
    """Configuration for verification checkpoints that nodal officers must complete"""
    checkpoint_key: str = Field(..., alias="checkpointKey", unique=True)
    label: str
    checkpoint_type: CheckpointType = Field(..., alias="checkpointType")
    description: Optional[str] = None
    is_required: bool = Field(default=True, alias="isRequired")
    is_enabled: bool = Field(default=True, alias="isEnabled")
    display_order: int = Field(default=0, alias="displayOrder")
    validation: Optional[Dict[str, Any]] = None  # e.g., {"min": 1, "max": 5} for rating
    options: Optional[List[str]] = None  # For dropdown type
    placeholder: Optional[str] = None
    hint: Optional[str] = None
    category: Optional[str] = None  # e.g., "credibility", "details", "facilities", "safety"
    created_at: datetime = Field(default_factory=datetime.utcnow, alias="createdAt")
    updated_at: datetime = Field(default_factory=datetime.utcnow, alias="updatedAt")

    @field_validator("checkpoint_type", mode="before")
    @classmethod
    def validate_checkpoint_type(cls, v):
        if isinstance(v, str):
            try:
                return CheckpointType(v)
            except ValueError:
                return v
        return v

    class Settings:
        name = "VerificationCheckpointConfig"

